src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs

```diff
@@ -65,6 +68,47 @@ public class SocketTransportOptions
         /// </remarks>
         public bool UnsafePreferInlineScheduling { get; set; }
 
+        /// <summary>
+        /// A function used to create a new <see cref="Socket"/> to listen with. If
+        /// not set, <see cref="CreateDefaultListenSocket" /> is called instead.
+        /// </summary>
+        public Func<EndPoint, Socket>? CreateListenSocket { get; set; }
+
+        /// <summary>
+        /// Creates a default instance of <see cref="Socket"/> for the given <see cref="EndPoint"/>
+        /// that can be used by a connection listener to listen for inbound requests.
+        /// </summary>
+        /// <param name="endpoint">
+        /// An <see cref="EndPoint"/>.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Socket"/> instance.
+        /// </returns>
+        public static Socket CreateDefaultListenSocket(EndPoint endpoint)
+        {
+            switch (endpoint)
+            {
+                case FileHandleEndPoint fileHandle:
+                    return new Socket(
+                        new SafeSocketHandle((IntPtr)fileHandle.FileHandle, ownsHandle: true)
```

Comment:
```suggestion
                        new SafeSocketHandle((IntPtr)fileHandle.FileHandle, ownsHandle: false)
```

Since we're not managing the lifetime anymore.

Reply:
Interestingly, when I do this the test case that uses a `FileHandleEndPoint` never returns - the shutdown of Kestrel hangs indefinitely (when calling `_transportManager.StopAsync()`) - presumably because the underlying socket isn't closed. In the case of the test this is because the underlying `Socket` I steal the handle from is static and I'm unable to manage its lifetime as it's consumed using `MemberData`.

I can absolutely fix the test case but I'm now dubious to change it because it's possible that Kestrel might hang if the caller doesn't clean-up the handle that they supposedly own. That seems like a bad place to be without much indication what the problem is.

Reply:
Thinking that adding a comment and leaving the `ownsHandle: true` might be a safer route to go here

Reply:
This makes me think we should just go the `Func<EndPoint, Func<EndPoint, Socket>, Socket>` route since it's the most-flexible/least-tricky way to do this while continuing to manage `_socketHandle` lifetime. This is a pretty advanced API anyway. I don't like the idea of SafeHandle's finalizer controlling the lifetime of the socket.

Reply:
Actually, it looks like Socket itself disposes the SafeSocketHandle if we claim ownership. It seems like we were doing a bunch of work for nothing. I think your change is good as is. Maybe just add a comment saying that `ownsHandle` indicates the [Socket itself will dispose the handle when it is Disposed](https://github.com/dotnet/runtime/blob/999c9c0bf1162e624533e2df58062d2811961fd3/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3270).